### PR TITLE
Engineering/Security Override Update

### PR DESCRIPTION
**Pull request by lessthnthree**
## About The Pull Request
The captain, heads, and AI now have greater control over the engineering override access system. It can now be toggled on/off on any alert level from the communications console or the keycard auth device, it's no longer automatically bound to orange alert. There may be gameplay reasons to turn on/off the access that may not warrant orange alert, and since orange alert is on par with blue security wise, if the station is on amber, red, or higher the captain (or heads via keycard auth) can override the system and turn the access on or off.

As per the new security policy, security staff receive similar expanded access on red alert and higher as engineers do on orange alert.

## How This Contributes To The Skyrat Roleplay Experience
- The station currently has no control over the engineering override system, this gives them that power to control and use it as they see fit.
- Adds the new expanded access for security.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>

https://user-images.githubusercontent.com/83487515/213295861-696a84fb-dbac-4545-89d5-048246126c62.mp4

https://user-images.githubusercontent.com/83487515/213295878-5581cd49-cd58-439d-96ae-a7a82bd69da6.mp4

</details>

## Changelog

:cl: LT3
qol: Engineering override access can now be manually controlled at the communications console and keycard auth devices
fix: Fixed missing undefine in airlock module
add: Security now has expanded airlock access on red alert and higher
/:cl:

### CODE REVIEW

File: code/game/machinery/computer/communications.dm [hunk 1]
```diff
@@ -571,6 +585,7 @@ GLOBAL_VAR_INIT(cops_arrived, FALSE)
 				data["canSendToSectors"] = FALSE
 				data["canSetAlertLevel"] = FALSE
 				data["canToggleEmergencyAccess"] = FALSE
+				data["canToggleEngineeringOverride"] = FALSE
```
Comment (Zonespace27) [inline code/game/machinery/computer/communications.dm]: SR edit comment

File: code/game/machinery/computer/communications.dm [hunk 2]
```diff
@@ -602,6 +617,8 @@ GLOBAL_VAR_INIT(cops_arrived, FALSE)
 				if (authenticated_as_silicon_or_captain(user))
 					data["canToggleEmergencyAccess"] = TRUE
 					data["emergencyAccess"] = GLOB.emergency_access
+					data["canToggleEngineeringOverride"] = TRUE
```
Comment (Zonespace27) [inline code/game/machinery/computer/communications.dm]: ditto

File: modular_skyrat/modules/engineering_override/code/engineering_override.dm [hunk 3]
```diff
@@ -1,5 +1,6 @@
 // This module sets airlocks in certain areas to be able to have an Engineer Override on orange alert.
 // Crew with ID cards with the engineering flag will be able to access these areas during those times.
+#define COMSIG_GLOB_FORCE_ENG_OVERRIDE "force_engineering_override"
```
Comment (Zonespace27) [inline modular_skyrat/modules/engineering_override/code/engineering_override.dm]: signal folder

File: modular_skyrat/modules/engineering_override/code/engineering_override.dm [hunk 4]
```diff
@@ -111,6 +119,43 @@
...
+		return
+
+/obj/machinery/door/airlock/proc/force_eng_override(datum/source, status)
+	SIGNAL_HANDLER
+
+	if(!status)
+		engineering_override = FALSE
+		normalspeed = TRUE
+		update_appearance()
+		return
```
Comment (Zonespace27) [inline modular_skyrat/modules/engineering_override/code/engineering_override.dm:143]:
```suggestion
		return
		
```

File: modular_skyrat/modules/engineering_override/code/engineering_override.dm [hunk 5]
```diff
@@ -111,6 +119,43 @@
...
+	var/area/source_area = get_area(src)
+	if(!source_area.engineering_override_eligible)
+		return
+
+	engineering_override = TRUE
+	normalspeed = FALSE
+	update_appearance()
+	return
```
Comment (Zonespace27) [inline modular_skyrat/modules/engineering_override/code/engineering_override.dm]: this return does nothing

File: tgui/packages/tgui/interfaces/CommunicationsConsole.js [hunk 6]
```diff
@@ -235,9 +235,11 @@ const PageMain = (props, context) => {
     canSendToSectors,
```
Comment (Zonespace27) [inline tgui/packages/tgui/interfaces/CommunicationsConsole.js:235]: none of these changes are documented with SR edit comments

File: tgui/packages/tgui/interfaces/KeycardAuth.js [hunk 7]
```diff
@@ -44,6 +44,12 @@ export const KeycardAuth = (props, context) => {
                       onClick={() => act('emergency_maint')}
                       content="Emergency Maintenance Access"
                     />
+                    <Button
```
Comment (Zonespace27) [inline tgui/packages/tgui/interfaces/KeycardAuth.js:48]: neither is this

File: modular_skyrat/modules/meteor_warning/code/meteor_warning.dm [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+	..()
+	start_when = rand(210,240)
+	end_when = start_when + METEOR_WAVE_DURATION
+
+/datum/round_event/meteor_wave/announce(fake)
+	priority_announce("Meteors have been detected on collision course with the station. The early warning system estimates first collision in approximately [start_when * 2] seconds. Ensure all sensitive areas and equipment are shielded.", "Meteor Alert", ANNOUNCER_METEORS)
+	var/level = SSsecurity_level.get_current_level_as_number()
+	addtimer(CALLBACK(src, PROC_REF(orange_alert)), 4 MINUTES, level)
+
+/datum/round_event/meteor_wave/proc/orange_alert(datum/round_event/mv)
```
Comment (Zonespace27) [inline modular_skyrat/modules/meteor_warning/code/meteor_warning.dm]: Let the station figure out if it needs orange or not, remove this

File: modular_skyrat/modules/meteor_warning/code/meteor_warning.dm [hunk 9]
```diff
@@ -0,0 +1,35 @@
+#define METEOR_WAVE_DURATION 37
+
+/datum/round_event/meteor_wave/New()
+	..()
+	start_when = rand(210,240)
```
Comment (Zonespace27) [inline modular_skyrat/modules/meteor_warning/code/meteor_warning.dm]: time defines

Comment (Zonespace27) [inline modular_skyrat/modules/meteor_warning/code/meteor_warning.dm]:
```suggestion
	start_when = rand(210, 240)
```

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Comment by ghost** [comment]
Two things:
- This should likely be two separate PRs
- 7-9 minutes feels like a long time, THOUGH considering you didn't change meteors to be directional I can understand it. Personally, if you'd be fine cutting it from your PR, I'd prefer to reopen my Directional Meteors/Early Warning port from oldbase, especially with TG actually having a directional framework now (I closed it while TG implemented that)

**Comment by lessthnthree** [comment]
> Two things:
> 
> * This should likely be two separate PRs
> * 7-9 minutes feels like a long time, THOUGH considering you didn't change meteors to be directional I can understand it. Personally, if you'd be fine cutting it from your PR, I'd prefer to reopen my Directional Meteors/Early Warning port from oldbase, especially with TG actually having a directional framework now (I closed it while TG implemented that)

Originally the two were tied together functionally, but now that orange alert isn't being fired by the event it can be split. How far away is your directional PR?

**Comment by ghost** [comment]
> 
> Originally the two were tied together functionally, but now that orange alert isn't being fired by the event it can be split. How far away is your directional PR?

Makes sense yeayea
Directional PR has been sitting dead in the water for a while, but seeing as TG added their own, it might be something you can even add to your own PR; I'll go comment on that one with how you'd likely do it, since you've already made it

File: modular_skyrat/modules/aesthetics/airlock/code/airlock.dm [hunk 10]
```diff
@@ -80,6 +83,10 @@
 				light_state = AIRLOCK_LIGHT_ENGINEERING
 				lights_overlay = "lights_engineering"
 				pre_light_color = light_color_engineering
+			else if(security_override)
+				light_state = AIRLOCK_LIGHT_SECURITY
+				lights_overlay = "lights_security"
+				pre_light_color = light_color_security
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/aesthetics/airlock/code/airlock.dm]: There's no reason to have it as a variable here.
```suggestion
				pre_light_color = AIRLOCK_SECURITY_LIGHT_COLOR
```

File: modular_skyrat/modules/aesthetics/airlock/code/airlock.dm [hunk 11]
```diff
@@ -40,6 +42,7 @@
 	var/light_color_access = AIRLOCK_ACCESS_LIGHT_COLOR
 	var/light_color_emergency = AIRLOCK_EMERGENCY_LIGHT_COLOR
 	var/light_color_engineering = AIRLOCK_ENGINEERING_LIGHT_COLOR
+	var/light_color_security = AIRLOCK_SECURITY_LIGHT_COLOR
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/aesthetics/airlock/code/airlock.dm]: Refer to my previous comment.
```suggestion
```

File: modular_skyrat/modules/aesthetics/airlock/code/airlock.dm [hunk 12]
```diff
@@ -560,12 +567,15 @@
...
+#undef AIRLOCK_LIGHT_ENGINEERING
+#undef AIRLOCK_LIGHT_SECURITY
+#undef AIRLOCK_ENGINEERING_LIGHT_COLOR
 #undef AIRLOCK_POWERON_LIGHT_COLOR
 #undef AIRLOCK_BOLTS_LIGHT_COLOR
 #undef AIRLOCK_ACCESS_LIGHT_COLOR
 #undef AIRLOCK_EMERGENCY_LIGHT_COLOR
 #undef AIRLOCK_DENY_LIGHT_COLOR
+#undef AIRLOCK_SECURITY_LIGHT_COLOR
 //SKYRAT EDIT END
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/aesthetics/airlock/code/airlock.dm:578]: While you're here, please remove the Skyrat Edit comments, it's literally a modular file.

File: modular_skyrat/modules/autoairlock_override/code/autoairlock.dm [hunk 13]
```diff
@@ -0,0 +1,45 @@
+/// Check for the four states of open access. Emergency, Unrestricted, Security, and Engineering Override
+/obj/machinery/door/airlock/allowed(mob/user)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/code/autoairlock.dm]: Rename this file (and module) to `auto_airlock` please, it's not something that would go in one word.

File: modular_skyrat/modules/autoairlock_override/code/engineering_override.dm [hunk 14]
```diff
@@ -9,6 +9,18 @@
...
+	engineering_override_eligible = TRUE
+
+/area/station/ai_monitored/turret_protected/aisat_interior
+	engineering_override_eligible = TRUE
+
+/area/station/ai_monitored/turret_protected/ai_upload
+	engineering_override_eligible = TRUE
+
+/area/station/ai_monitored/turret_protected/ai_upload_foyer
+	engineering_override_eligible = TRUE
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/code/engineering_override.dm]: I feel like these aren't areas we'd want Engineers strolling about actually.

File: modular_skyrat/modules/autoairlock_override/code/engineering_override.dm [hunk 15]
```diff
@@ -111,17 +126,38 @@
...
-				airlock.do_animate("deny")
-				if(airlock.emergency)
-					airlock.emergency = FALSE
-					airlock.update_appearance()
-				if(airlock.engineering_override)
-					airlock.engineering_override = FALSE
-					airlock.update_appearance()
+// Manual override for when it's not orange alert.
+GLOBAL_VAR_INIT(force_eng_override, FALSE)
+/proc/toggle_eng_override()
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/code/engineering_override.dm]: Also, add documentation to this proc.
```suggestion

/proc/toggle_eng_override()
```

File: modular_skyrat/modules/autoairlock_override/code/engineering_override.dm [hunk 16]
```diff
@@ -111,17 +126,38 @@
...
+		return
+	else
+		GLOB.force_eng_override = FALSE
+		minor_announce("Expanded engineering access has been revoked.", "Engineering Emergency")
+		var/level = SSsecurity_level.get_current_level_as_number()
+		SSblackbox.record_feedback("nested tally", "keycard_auths", 1, list("engineer override access", "disabled"))
+		if(level == SEC_LEVEL_ORANGE)
+			return
+		SEND_GLOBAL_SIGNAL(COMSIG_GLOB_FORCE_ENG_OVERRIDE, FALSE)
+		return
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/code/engineering_override.dm]: This also just feels like it's a lot of repetition of code that could be done in more concise way.
```suggestion

	GLOB.force_eng_override = FALSE
	minor_announce("Expanded engineering access has been revoked.", "Engineering Emergency")
	var/level = SSsecurity_level.get_current_level_as_number()
	SSblackbox.record_feedback("nested tally", "keycard_auths", 1, list("engineer override access", "disabled"))
	if(level == SEC_LEVEL_ORANGE)
		return

	SEND_GLOBAL_SIGNAL(COMSIG_GLOB_FORCE_ENG_OVERRIDE, FALSE)
```

File: modular_skyrat/modules/autoairlock_override/code/engineering_override.dm [hunk 17]
```diff
@@ -111,17 +126,38 @@
...
+
+/obj/machinery/door/airlock/proc/force_eng_override(datum/source, status)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/code/engineering_override.dm]: Document this proc please.

File: modular_skyrat/modules/autoairlock_override/code/security_override.dm [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+		update_appearance()
+		return
+
+	security_override = FALSE
+	update_appearance()
+	return
+
+// Manual override for when it's not red alert.
+GLOBAL_VAR_INIT(force_sec_override, FALSE)
+/proc/toggle_sec_override()
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/code/security_override.dm]: Document this proc please.

File: modular_skyrat/modules/autoairlock_override/code/security_override.dm [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+	if(level >= SEC_LEVEL_RED)
+		security_override = TRUE
+		update_appearance()
+		return
+
+	security_override = FALSE
+	update_appearance()
+	return
+
+// Manual override for when it's not red alert.
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/code/security_override.dm]: Also for these it's `/// ` not just `// `

File: modular_skyrat/modules/autoairlock_override/code/security_override.dm [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+		return
+	else
+		GLOB.force_eng_override = FALSE
+		minor_announce("Expanded security access has been disabled.", "Security Emergency")
+		var/level = SSsecurity_level.get_current_level_as_number()
+		SSblackbox.record_feedback("nested tally", "keycard_auths", 1, list("security override access", "disabled"))
+		if(level >= SEC_LEVEL_RED)
+			return
+		SEND_GLOBAL_SIGNAL(COMSIG_GLOB_FORCE_SEC_OVERRIDE, FALSE)
+		return
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/code/security_override.dm]:
```suggestion

	GLOB.force_sec_override = FALSE
	minor_announce("Expanded security access has been disabled.", "Security Emergency")
	var/level = SSsecurity_level.get_current_level_as_number()
	SSblackbox.record_feedback("nested tally", "keycard_auths", 1, list("security override access", "disabled"))
	if(level >= SEC_LEVEL_RED)
		return

	SEND_GLOBAL_SIGNAL(COMSIG_GLOB_FORCE_SEC_OVERRIDE, FALSE)
```

File: modular_skyrat/modules/autoairlock_override/code/security_override.dm [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+
+	security_override = FALSE
+	update_appearance()
+	return
+
+// Manual override for when it's not red alert.
+GLOBAL_VAR_INIT(force_sec_override, FALSE)
+/proc/toggle_sec_override()
+	if(!GLOB.force_sec_override)
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/code/security_override.dm]: You should probably change the value of this before the `if()` so it's less confusing.

File: modular_skyrat/modules/autoairlock_override/readme.md [hunk 22]
```diff
@@ -2,17 +2,24 @@ https://github.com/Skyrat-SS13/Skyrat-tg/pull/16787
 
 ## Title:
 Engineering override for airlocks on orange alert
-MODULE ID: ENGINEERING_OVERRIDE
+MODULE ID: AUTOAIRLOCK_OVERRIDE
```
Comment (GoldenAlpharex) [inline modular_skyrat/modules/autoairlock_override/readme.md]:
```suggestion
MODULE ID: AUTO_AIRLOCK_OVERRIDE
```

File: tgui/packages/tgui/interfaces/CommunicationsConsole.js [hunk 23]
```diff
@@ -235,9 +235,11 @@ const PageMain = (props, context) => {
...
     canSetAlertLevel,
     canToggleEmergencyAccess,
+    canToggleEngineeringOverride, // SKYRAT EDIT - Engineering Override
     emagged,
     syndicate,
     emergencyAccess,
+    engineeringOverride, // SKYRAT EDIT - Engineering Override
```
Comment (GoldenAlpharex) [inline tgui/packages/tgui/interfaces/CommunicationsConsole.js:242]: Wouldn't you want to have the security override stuff here too?

File: tgui/packages/tgui/interfaces/CommunicationsConsole.js [hunk 24]
```diff
@@ -382,15 +384,26 @@ const PageMain = (props, context) => {
           )}
           {!!canToggleEmergencyAccess && (
             <Button.Confirm
-              icon="id-card-o"
+              icon="wrench" // SKYRAT EDIT
```
Comment (GoldenAlpharex) [inline tgui/packages/tgui/interfaces/CommunicationsConsole.js]: Leave this one as it is and put the wrench on the engineering one, that makes a lot more sense.
```suggestion
              icon="id-card-o"
```

File: tgui/packages/tgui/interfaces/CommunicationsConsole.js [hunk 25]
```diff
@@ -382,15 +384,26 @@ const PageMain = (props, context) => {
...
               } Emergency Maintenance Access`}
               color={emergencyAccess ? 'bad' : undefined}
               onClick={() => act('toggleEmergencyAccess')}
             />
           )}
-
+          {/* SKYRAT EDIT ADDITION START */}
+          {!!canToggleEngineeringOverride && (
+            <Button.Confirm
+              icon="id-card-o"
```
Comment (GoldenAlpharex) [inline tgui/packages/tgui/interfaces/CommunicationsConsole.js]:
```suggestion
              icon=""wrench"
```

File: tgui/packages/tgui/interfaces/CommunicationsConsole.js [hunk 26]
```diff
@@ -382,15 +384,26 @@ const PageMain = (props, context) => {
...
+              icon="wrench" // SKYRAT EDIT
               content={`${
                 emergencyAccess ? 'Disable' : 'Enable'
               } Emergency Maintenance Access`}
               color={emergencyAccess ? 'bad' : undefined}
               onClick={() => act('toggleEmergencyAccess')}
             />
           )}
-
+          {/* SKYRAT EDIT ADDITION START */}
```
Comment (GoldenAlpharex) [inline tgui/packages/tgui/interfaces/CommunicationsConsole.js]:
```suggestion
          {/* SKYRAT EDIT ADDITION START - Engineering Override */}
```

File: tgui/packages/tgui/interfaces/KeycardAuth.js [hunk 27]
```diff
@@ -58,6 +67,7 @@ export const KeycardAuth = (props, context) => {
                         content="Permit-Locked Firing Pin Unrestriction"
                       />
                     )}
+                    {/* SKYRAT EDIT ADDITION EMD */}
```
Comment (GoldenAlpharex) [inline tgui/packages/tgui/interfaces/KeycardAuth.js]:
```suggestion
                    {/* SKYRAT EDIT ADDITION END */}
```

File: tgui/packages/tgui/interfaces/KeycardAuth.js [hunk 28]
```diff
@@ -44,12 +44,21 @@ export const KeycardAuth = (props, context) => {
...
+                      content="Engineering Override Access"
+                    />
+                    {/* SKYRAT EDIT ADDITION END */}
                     <Button
                       icon="meteor"
                       fluid
                       onClick={() => act('bsa_unlock')}
                       content="Bluespace Artillery Unlock"
                     />
+                    {/* SKYRAT EDIT ADDITION START */}
```
Comment (GoldenAlpharex) [inline tgui/packages/tgui/interfaces/KeycardAuth.js]:
```suggestion
                    {/* SKYRAT EDIT ADDITION START - Permit Pins */}
```

**Review by GoldenAlpharex — CHANGES_REQUESTED** [review]
I haven't checked the sprites yet, and it's probably something that'll be mostly for @Ebin-Halcyon to give his approval to.

**Comment by ghost** [comment]
The icon changes in this are a bit of a mess, several airlock states are changed, the external one is partially recolored but not fully in the animation, and lots of the extra stripes feel redundant and, moreover, aren't shown in the PR.

I recommend the manual engineering override stay its own PR for your own sake, you're kind of over-extending. The security access being a separate PR will make it easier to review code-wise, icon-wise, and for you to keep changes documented

**Comment by lessthnthree** [comment]
Yeah I'll split it in two.

**Pull request closed without merging**